indexing-service/src/test/java/io/druid/indexing/common/task/TaskContextTest.java

```diff
@@ -0,0 +1,53 @@
+/*
+* Licensed to Metamarkets Group Inc. (Metamarkets) under one
+* or more contributor license agreements. See the NOTICE file
+* distributed with this work for additional information
+* regarding copyright ownership. Metamarkets licenses this file
+* to you under the Apache License, Version 2.0 (the
+* "License"); you may not use this file except in compliance
+* with the License. You may obtain a copy of the License at
+*
+* http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing,
+* software distributed under the License is distributed on an
+* "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+* KIND, either express or implied. See the License for the
+* specific language governing permissions and limitations
+* under the License.
+*/
+
+package io.druid.indexing.common.task;
+
+import com.google.common.collect.ImmutableMap;
+import junit.framework.Assert;
+import org.junit.Test;
+
+public class TaskContextTest
+{
+  public interface IContext
+  {
+
+  }
+
+  public static class Impl1 implements IContext
+  {
+
+  }
+
+  public static class Impl2 implements IContext
+  {
+
+  }
+
+  @Test
+  public void testgetContextWithDefaultValue()
+  {
+    IContext contextValue = new Impl1();
+    NoopTask noopTask = new NoopTask("test", 0l, 0l, "YES", null, ImmutableMap.<String, Object>of("k1", contextValue));
+    Assert.assertEquals(contextValue, noopTask.getContextValue("k1"));
+    // Verify Typecaset by passing object of another type as Default value
+    IContext val = noopTask.getContextValue("k1", new Impl2());
+    Assert.assertEquals(contextValue, val);
+  }
```

Comment:
java is funny with generics.
Can you try context = ImmutableMap.<String, Object>of("k1", 10)
then try getting getContextValue("k1","hello")

for me it failed and I belive to make it really work you have to do what @drcrallen suggested. However, at the same time, I don't know if need so much complication is really needed here. Is getContextValue(key,default) really essential or we should just remove it.
For me getContext() and getContextValue(key) would already be enough.